### PR TITLE
enable unlimited send of marketing emails

### PR DESCRIPTION
**Pull request by regevbr**
| Q  | A
| --- | ---
| Bug fix? |  No
| New feature? |  Yes
| Automated tests included? | No
| Related user documentation PR URL | WIP
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | N/A
| BC breaks? |  No
| Deprecations? |  No

#### Description:

Add the ability to send marketing emails unlimited number of times.
Currently, there is no way to send the same email many times, while controlling the email frequency. This feature enables it.
Transactional emails are sent regardless of frequency, and marketing emails follow the frequency rules, but can only be sent once per lead.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Enable unlimited marketing emails s in the email configuration section
3. Send a marketing email to lead, and then send it again and make sure both or sent

### CODE REVIEW

**Comment by AlbanL74fr** [comment]
Hi @regevbr Can you elaborate the use case? Is your email sent from a campaign? 
If yes, and if your template of email is sent as "Marketing" in campaign 1, what prevents you to send the same template as "Transactional" in campaign 2?

**Comment by regevbr** [comment]
Our use case is as follows:

1. We have a single campaign which a user can enter multiple times.
2. We have an abandoned cart email we want to send via that campaign (and not any other campaign).
3. We don't wan't to send emails to customers more than once a day.
4. We do want to be able to send that email multiple times.

If we set the email to be transnational, it will be sent every time - violating `#3`
If we set the email to be marketing, it will only be sent once, and never again - violating `#4`

With my feature enabled, the use case works perfectly

@AlbanL74fr does that make sense?

**Comment by AlbanL74fr** [comment]
I understand better, @regevbr. The use case is clear. For better UX, I would suggest use Transactional (as you want the email to be send over and ovee again) and add a check box i.e "Force apply frequency limit". Imagine an input field shows and you set a number i.e. "1", and choose period as a select box i.e. "day".
It would be more intuitive.
Not being a dev, I am not sure if that is possible.

**Comment by regevbr** [comment]
@AlbanL74fr but I do still want to keep the frequency control on the global level and not specific email... Your suggestion will cause the user to get more than 1 email a day if I have other campgins running. What do you think?

**Review by Strubbel94 — DISMISSED** [review]
This feature is working fine.
no problems.

File: app/bundles/EmailBundle/EventListener/CampaignSubscriber.php [hunk 1]
```diff
@@ -73,21 +74,24 @@ class CampaignSubscriber implements EventSubscriberInterface
...
      */
     public function __construct(
         LeadModel $leadModel,
         EmailModel $emailModel,
         EventModel $eventModel,
         MessageQueueModel $messageQueueModel,
         SendEmailToUser $sendEmailToUser,
-        TranslatorInterface $translator
+        TranslatorInterface $translator,
+        MauticFactory $factory
```
Comment (kuzmany) [inline app/bundles/EmailBundle/EventListener/CampaignSubscriber.php]: MauticFactory is deprecated and will remove in 3.0
Use service, in your case CoreParametersHelper

**Review by kuzmany — DISMISSED** [review]

File: app/bundles/EmailBundle/EventListener/CampaignSubscriber.php (hunk 1 above)
Comment (regevbr) [reply app/bundles/EmailBundle/EventListener/CampaignSubscriber.php]: @kuzmany thanks for the feedback!
I changed the code, please review it again.

**Comment by kuzmany** [comment]
I think for this purpose is better use transactional emails and remove from campaign you can delayed as long as you want (this prevent go through campaign more times)

**Comment by regevbr** [comment]
@kuzmany But I do want leads to enter the campaign multiple times...

**Comment by kuzmany** [comment]
@regevbr 
your wrote

> We have a single campaign which a user can enter multiple times.
> We have an abandoned cart email we want to send via that campaign (and not any other campaign).

Abandoned cart do the same. 

- Contact added to campaign, send email, remove from campaign

In another time, same situation.

**Comment by regevbr** [comment]
@kuzmany But then I will by pass the global rate limits...

**Comment by regevbr** [comment]
@kuzmany maybe I don't understand the flow correctly, But I don't understand how your suggestions solves:

> Our use case is as follows:
> 
> 1. We have a single campaign which a user can enter multiple times.
> 2. We have an abandoned cart email we want to send via that campaign (and not any other campaign).
> 3. We don't wan't to send emails to customers more than once a day.
> 4. We do want to be able to send that email multiple times.
> 
> If we set the email to be transnational, it will be sent every time - violating `#3`
> If we set the email to be marketing, it will only be sent once, and never again - violating `#4`

**Comment by kuzmany** [comment]
@regevbr but how you are able to send email multiple times by campaigns? Action is triggered just one time, and then you need repeat the campaign with remove contact from campaign. I suggest delay this remove from campaign

**Comment by regevbr** [comment]
@kuzmany Every leaf in the decision tree is an exit action from the campaign, making the user being able to start it again.
If I add a delay before the exit action, it will work as you suggest, but then I will have to send transnational email, which overides the global rate limiting of emails per lead...

**Comment by kuzmany** [comment]
I check the code, and I understand now. Yes, good idea.
But I think at the moment wording and UX is not clear for me.
Somebody from Mautic core team should put opinion on it @escopecz

**Comment by escopecz** [comment]
A screenshot with the new field would make this discussion more to the point. I can imagine how it looks like based on the code diff, but not everyone can read the code.

I also have an issue with UX. An abandoned cart email belongs to the transactional email category in my mind. This seems like a hack of marketing emails for your specific use case. I'm sure others would like to use such feature but it will be hard to explain why the marketers must select marketing email type for abandoned cart email.

I'm not an expert on UI/UX by any means. @dbhurley would you like to share your opinion?

**Comment by kuzmany** [comment]
The point is marketing messages with limits (daily, weekly) without limit for on time on each email.

But how it provide clearly?

**Comment by regevbr** [comment]
I attached a screenshot.
I'm the farthest thing away from UI/UX and I'm sure the phrasing here should be changed, just let me know what to do.

image

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
Hi @regevbr are you still able to work on this PR? It has outstanding conflicts and we haven't heard from you about whether it's still relevant in Mautic 3.x. We also would need you to sign the Contributors Agreement (see comment above for the link).

I believe that the abandoned cart email would come under a transactional rather than marketing email so I think that there are some fundamental issues with the proposal in this PR that we should discuss more widely.

If you can get back to us and let us know if you're still able to contribute then we can discuss how to move forward. If we don't hear from you within 14 days we'll close the PR.

**Comment by npracht** [comment]
I'll close this PR as far as there is no feedbacks about the owner. Don't take it wrong, part of it is our fault, we've been too long (on the Mautic Community Product Team side) to handle it at the time the PR has been offer.

If someone want to take it, i'll reopen it or they can create a new PR.

**Pull request closed without merging**